File: solutions/typescript/034-trapping-rain-water.ts
/**
 * Problem 34: Trapping Rain Water (LeetCode 42)
 * Difficulty: Hard
 * Language: TypeScript
 */
function trap(height: number[]): number {
  const maxLeft = height.reduce<number[]>(
    (acc, h) => [...acc, Math.max(h, acc.length ? acc[acc.length - 1] : 0)], []);
  const maxRight = height.reduceRight<number[]>(
    (acc, h) => [Math.max(h, acc.length ? acc[0] : 0), ...acc], []);
  return height.reduce(
    (sum, h, i) => sum + Math.max(0, Math.min(maxLeft[i], maxRight[i]) - h), 0);
}

File: solutions/clojure/034-trapping-rain-water.clj
;; Problem 34: Trapping Rain Water (LeetCode 42)
;; Difficulty: Hard
;; Language: Clojure
;; 
(defn trap [height]
  (let [max-left  (reductions max height)
        max-right (reverse (reductions max (reverse height)))]
    (reduce + (map (fn [h l r] (max 0 (- (min l r) h)))
                   height max-left max-right))))
